ulb.py

```diff
@@ -97,31 +140,97 @@ class Data(ct.Structure):
+dump_map()
+print("... condig applied to bpf.")
+
+# started
+print("\nLoad balancing UDP traffic over {} interface for port(s) {} from :".format(ifnet, ports, ip_ntostr(virtual_server['ip'])))
 for real_server in real_servers:
-    real_servers_array[real_servers_array.Key(i)] = real_servers_array.Leaf(real_server['ip'])
-    real_servers_map[real_servers_map.Key(real_server['ip'])] = real_servers_map.Leaf(real_server['ip'])
-    i+=1
+    print ("VIP:{} <=======> Real Server:{}".format(server_tostr(virtual_server), server_tostr(real_server)))
 
-# Utility function to print udp dest NAT.
+# Utility function to print udp NAT.
 def print_event(cpu, data, size):
     event = ct.cast(data, ct.POINTER(Data)).contents
     print("source {} --> dest {} {}".format(ip_mac_tostr(event.smac, event.saddr),ip_mac_tostr(event.dmac, event.daddr), associationType_tostr(event.associationType)))
 
-# Loop to read perf buffer
-b["events"].open_perf_buffer(print_event)
-while 1:
-    try:
-        b.perf_buffer_poll()
-        # DEBUG STUFF
-        #(task, pid, cpu, flags, ts, msg) = b.trace_fields(nonblocking = True)
-        #while msg:
-        #    print("%s \n" % (msg))
-        #    (task, pid, cpu, flags, ts, msg) = b.trace_fields(nonblocking = True)
-    except ValueError:
-        continue
-    except KeyboardInterrupt:
-        break;
-
-# Detach bpf progam
-b.remove_xdp(ifnet)
+# Program loop
+try:
+    b["events"].open_perf_buffer(print_event)
+    while 1:
+        try:
+            # read and log perf_buffer
+            b.perf_buffer_poll(1000)
+            # watch if config file changed
+            new_mtime = os.stat(config_file.name).st_mtime
+            if  new_mtime != config_file_mtime:
+                # load real server from config
+                new_real_servers = None
+                try:
+                    config_file_mtime = new_mtime
+                    with open(config_file.name) as f:
+                        new_real_servers = load_config(f)
+                except Exception as e:
+                    print ("Unable to load config {} file : {}".format(config_file.name, e))
+                    print ("Old Config is keeping : {}".format(servers_tostr(real_servers)))
+
+                # if succeed try to update bpf map
+                if new_real_servers is not None:
+                    print("Apply new config ...")
+                    update_real_server(real_servers, new_real_servers)
+                    real_servers = new_real_servers
+                    dump_map()
+                    print("... new config applied.")
+            # DEBUG STUFF
+            #(task, pid, cpu, flags, ts, msg) = b.trace_fields(nonblocking = True)
+            #while msg:
+            #    print("%s \n" % (msg))
+            #    (task, pid, cpu, flags, ts, msg) = b.trace_fields(nonblocking = True)
+        except KeyboardInterrupt:
+            break;
+finally:
+    # Detach bpf progam
+    b.remove_xdp(ifnet)
```

Comment:
Will the XDP also be removed if the process receives a SIGTERM? That's what systemd sends to stop a service.

Reply:
Good question I can test this.

Reply:
Tested with `kill -SIGTERM PID` and looking at `ip link`
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 xdpgeneric qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    prog/xdp id 48 
```

XDP prog is still attached (=> prog/xdp id 48)

So I will create a PR for this.

Is SIGTERM the only one SIGNAL we want to handle ?

Reply:
Yes, SIGTERM and SIGINT are the most important to handle.